File: desktop-tauri/src/ipc/accountBilling.ts
import { invoke } from "@tauri-apps/api/core";

import type { CreditsSummary, TopupOption } from "../types";

/** Membership and credits. Every address these open is chosen natively: a
 * one-time Stripe URL never crosses into the renderer. */

export function accountCredits(): Promise<CreditsSummary> {
  return invoke("account_credits");
}

export function accountTopupOptions(): Promise<TopupOption[]> {
  return invoke("account_topup_options");
}

export function accountBillingPortal(): Promise<void> {
  return invoke("account_billing_portal");
}

export function accountBillingTopup(topup: string): Promise<void> {
  return invoke("account_billing_topup", { topup });
}

/** Opens an enumerated page: the website's plans page, or Apple's
 * subscriptions page. The renderer names a page, never a URL. */
export function accountBillingPage(page: "plans" | "appStore"): Promise<void> {
  return invoke("account_billing_page", { page });
}

File: desktop-tauri/src-tauri/src/commands/account_billing.rs
use tauri::State;

use crate::account_billing::{self, CreditsSummary, TopupOption};
use crate::state::AppState;

#[tauri::command]
pub async fn account_credits(state: State<'_, AppState>) -> Result<CreditsSummary, String> {
    account_billing::credits(&state).await
}

#[tauri::command]
pub async fn account_topup_options() -> Result<Vec<TopupOption>, String> {
    account_billing::topups().await
}

/// Opens the Stripe customer portal in the system browser.
#[tauri::command]
pub async fn account_billing_portal(state: State<'_, AppState>) -> Result<(), String> {
    account_billing::open_portal(&state).await
}

/// Opens Stripe Checkout for one top-up size.
#[tauri::command]
pub async fn account_billing_topup(
    state: State<'_, AppState>,
    topup: String,
) -> Result<(), String> {
    account_billing::open_topup_checkout(&state, topup).await
}

/// Opens an enumerated billing page: `plans` or `appStore`. The renderer
/// names a page, never a URL.
#[tauri::command]
pub fn account_billing_page(page: String) -> Result<(), String> {
    account_billing::open_page(&page)
}
